xla/service/cpu/onednn_contraction_rewriter.cc

```diff
@@ -470,6 +477,242 @@ bool OneDnnContractionRewriter::ShouldRewriteConv(
+  // We need to add bitcast and copy instructions that were in the original
+  // pattern after each operand. It is possible that bitcast is not followed
+  // by copy, we will handle that inside create_bitcast_copy also.
+  if (copy_src != nullptr || bitcast_src != nullptr) {
+    create_bitcast_copy(quant_src, bitcast_src, copy_src, new_bitcast_src,
+                        *new_copy_src);
+  } else {
+    *new_copy_src = quant_src;
+  }
+  if (copy_wei != nullptr || bitcast_wei != nullptr) {
+    create_bitcast_copy(quant_wei, bitcast_wei, copy_wei, new_bitcast_wei,
+                        *new_copy_wei);
+  } else {
+    *new_copy_wei = quant_wei;
+  }
+
+  const float scale = 1.0;
+  const int zp = 0;
+  auto set_default_value = [&]<typename T>(HloInstruction*& instr, T value) {
+    instr = contraction_instr->AddInstruction(
+        HloInstruction::CreateConstant(LiteralUtil::CreateR0<T>(value)));
+  };
+  // If scale and/or zp are not found. That means they had a default values
+  // that were optimized out. Hence, we set a default value of 1.0 and 0 for
+  // scale and zp respectively.
+  if (*src_scale == nullptr) {
+    set_default_value(*src_scale, scale);
+  }
+  if (*src_zp == nullptr) {
+    set_default_value(*src_zp, zp);
+  }
+  if (*wei_scale == nullptr) {
+    set_default_value(*wei_scale, scale);
+  }
+  if (*wei_zp == nullptr) {
+    set_default_value(*wei_zp, zp);
+  }
+}
+
+bool ValidateQuantizationParams(HloInstruction* src_scale,
+                                HloInstruction* src_zp,
+                                HloInstruction* wei_scale,
+                                HloInstruction* wei_zp) {
+  if (!src_scale || !src_zp || !wei_zp) return false;
+  // oneDNN only supports scalar scale and zero-point for src and dst.
+  if (!IsScalar(src_scale) || !IsScalar(src_zp)) return false;
+
+  // Weights scale can only be scalar or vector
+  if (!IsScalar(wei_scale) && !IsVector(wei_scale)) return false;
+
+  // oneDNN only supports zero-point=0 for weights.
+  auto wei_zp_value = GetScalarConstantValue(wei_zp);
+  if (!wei_zp_value || wei_zp_value.value() != 0) return false;
+
+  return true;
+}
+
+// Check if the Quantize and Dequantize pairs have the same scale and zero point
+// values.
+bool ValidateQDQParams(HloInstruction* dequant_input,
+                       HloInstruction* dequant_scale,
+                       HloInstruction* dequant_zp) {
+  HloInstruction *quant_zp = nullptr, *quant_scale = nullptr, *input = nullptr,
+                 *clamp_min = nullptr, *clamp_max = nullptr;
+
+  bool quant_pattern = Match(
+      dequant_input,
+      QuantizePattern(&quant_scale, &quant_zp, &input, &clamp_min, &clamp_max));
+
+  if (quant_scale && dequant_scale) {
+    auto quant_scale_value = GetScalarConstantValue(quant_scale);
+    auto dequant_scale_value = GetScalarConstantValue(dequant_scale);
+    if (quant_scale_value && dequant_scale_value) {
```

Comment:
Should we consder the case where one of [quant_scale_value, dequant_scale_value] is null and other not?

same for the zp check below.

Reply:
yes, I did some refactoring to properly check for scale/zp matching.